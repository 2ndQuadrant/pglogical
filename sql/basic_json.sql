\i sql/basic_setup.sql

-- Simple decode with text-format tuples
SELECT data
FROM pg_logical_slot_peek_changes('regression_slot',
	NULL, NULL,
	'expected_encoding', 'UTF8',
	'min_proto_version', '1',
	'max_proto_version', '1',
	'startup_params_format', '1',
	'proto_format', 'json',
	'no_txinfo', 't');

\i sql/basic_teardown.sql
