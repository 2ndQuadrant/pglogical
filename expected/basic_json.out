\i sql/basic_setup.sql
SET synchronous_commit = on;
-- Schema setup
CREATE TABLE demo (
	seq serial primary key,
	tx text,
	ts timestamp,
	jsb jsonb,
	js json,
	ba bytea
);
SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'pglogical_output');
 ?column? 
----------
 init
(1 row)

-- Queue up some work to decode with a variety of types
INSERT INTO demo(tx) VALUES ('textval');
INSERT INTO demo(ba) VALUES (BYTEA '\xDEADBEEF0001');
INSERT INTO demo(ts, tx) VALUES (TIMESTAMP '2045-09-12 12:34:56.00', 'blah');
INSERT INTO demo(js, jsb) VALUES ('{"key":"value"}', '{"key":"value"}');
-- Rolled back txn
BEGIN;
DELETE FROM demo;
INSERT INTO demo(tx) VALUES ('blahblah');
ROLLBACK;
-- Multi-statement transaction with subxacts
BEGIN;
SAVEPOINT sp1;
INSERT INTO demo(tx) VALUES ('row1');
RELEASE SAVEPOINT sp1;
SAVEPOINT sp2;
UPDATE demo SET tx = 'update-rollback' WHERE tx = 'row1';
ROLLBACK TO SAVEPOINT sp2;
SAVEPOINT sp3;
INSERT INTO demo(tx) VALUES ('row2');
INSERT INTO demo(tx) VALUES ('row3');
RELEASE SAVEPOINT sp3;
SAVEPOINT sp4;
DELETE FROM demo WHERE tx = 'row2';
RELEASE SAVEPOINT sp4;
SAVEPOINT sp5;
UPDATE demo SET tx = 'updated' WHERE tx = 'row1';
COMMIT;
-- txn with catalog changes
BEGIN;
CREATE TABLE cat_test(id integer);
INSERT INTO cat_test(id) VALUES (42);
COMMIT;
-- Aborted subxact with catalog changes
BEGIN;
INSERT INTO demo(tx) VALUES ('1');
SAVEPOINT sp1;
ALTER TABLE demo DROP COLUMN tx;
ROLLBACK TO SAVEPOINT sp1;
INSERT INTO demo(tx) VALUES ('2');
COMMIT;
-- Simple decode with text-format tuples
SELECT data
FROM pg_logical_slot_peek_changes('regression_slot',
	NULL, NULL,
	'expected_encoding', 'UTF8',
	'min_proto_version', '1',
	'max_proto_version', '1',
	'startup_params_format', '1',
	'proto_format', 'json',
	'no_txinfo', 't');
                                                                                                                                                                                                                                                                                                                                                                                          data                                                                                                                                                                                                                                                                                                                                                                                           
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {"action":"S", "params": {"max_proto_version":"1","min_proto_version":"1","coltypes":"f","pg_version_num":"90600","pg_version":"9.6devel","pg_catversion":"201510161","database_encoding":"UTF8","encoding":"UTF8","forward_changesets":"f","forward_changeset_origins":"f","binary.internal_basetypes":"f","binary.binary_basetypes":"f","binary.basetypes_major_version":"906","binary.sizeof_int":"4","binary.sizeof_long":"8","binary.sizeof_datum":"8","binary.maxalign":"8","binary.bigendian":"f","binary.float4_byval":"t","binary.float8_byval":"t","binary.integer_datetimes":"t","binary.binary_pg_version":"906","no_txinfo":"t","hooks.startup_hook_enabled":"f","hooks.shutdown_hook_enabled":"f","hooks.row_filter_enabled":"f","hooks.transaction_filter_enabled":"f"}}
 {"action":"B", has_catalog_changes:"f"}
 {"action":"I","relation":["public","demo"],"newtuple":{"seq":1,"tx":"textval","ts":null,"jsb":null,"js":null,"ba":null}}
 {{"action":"C"}}
 {"action":"B", has_catalog_changes:"f"}
 {"action":"I","relation":["public","demo"],"newtuple":{"seq":2,"tx":null,"ts":null,"jsb":null,"js":null,"ba":"\\xdeadbeef0001"}}
 {{"action":"C"}}
 {"action":"B", has_catalog_changes:"f"}
 {"action":"I","relation":["public","demo"],"newtuple":{"seq":3,"tx":"blah","ts":"2045-09-12T12:34:56","jsb":null,"js":null,"ba":null}}
 {{"action":"C"}}
 {"action":"B", has_catalog_changes:"f"}
 {"action":"I","relation":["public","demo"],"newtuple":{"seq":4,"tx":null,"ts":null,"jsb":{"key": "value"},"js":{"key":"value"},"ba":null}}
 {{"action":"C"}}
 {"action":"B", has_catalog_changes:"f"}
 {"action":"I","relation":["public","demo"],"newtuple":{"seq":6,"tx":"row1","ts":null,"jsb":null,"js":null,"ba":null}}
 {"action":"I","relation":["public","demo"],"newtuple":{"seq":7,"tx":"row2","ts":null,"jsb":null,"js":null,"ba":null}}
 {"action":"I","relation":["public","demo"],"newtuple":{"seq":8,"tx":"row3","ts":null,"jsb":null,"js":null,"ba":null}}
 {"action":"D","relation":["public","demo"],"oldtuple":{"seq":7,"tx":null,"ts":null,"jsb":null,"js":null,"ba":null}}
 {"action":"U","relation":["public","demo"],"newtuple":{"seq":6,"tx":"updated","ts":null,"jsb":null,"js":null,"ba":null}}
 {{"action":"C"}}
 {"action":"B", has_catalog_changes:"t"}
 {"action":"I","relation":["public","cat_test"],"newtuple":{"id":42}}
 {{"action":"C"}}
 {"action":"B", has_catalog_changes:"f"}
 {"action":"I","relation":["public","demo"],"newtuple":{"seq":9,"tx":"1","ts":null,"jsb":null,"js":null,"ba":null}}
 {"action":"I","relation":["public","demo"],"newtuple":{"seq":10,"tx":"2","ts":null,"jsb":null,"js":null,"ba":null}}
 {{"action":"C"}}
(27 rows)

\i sql/basic_teardown.sql
SELECT 'drop' FROM pg_drop_replication_slot('regression_slot');
 ?column? 
----------
 drop
(1 row)

DROP TABLE demo;
DROP TABLE cat_test;
